File: src/AutoRest.SdkExplorer/Model/Code/ApiDesc.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

#nullable disable
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using AutoRest.SdkExplorer.Model.Azure;
using AutoRest.SdkExplorer.Model.Example;
using AutoRest.SdkExplorer.Model.Schema;
using AutoRest.SdkExplorer.Model.Hint;
using AutoRest.SdkExplorer.Utilities;

namespace AutoRest.SdkExplorer.Model.Code
{
    public class ApiDesc
    {
        public class ProviderType
        {
            public const string Resource = nameof(ProviderType.Resource);
            public const string ResourceCollection = nameof(ProviderType.ResourceCollection);
            public const string Extension = nameof(ProviderType.Extension);
        };

        public string Language { get; set; }
        public string SdkPackageName { get; set; }
        public string SdkPackageVersion { get; set; }
        public string ExplorerCodeGenVersion { get; set; }
        public string GeneratedTimestamp { get; set; }
        public List<string> Dependencies { get; set; } = new List<string>();

        public string ServiceName { get; set; }
        public string ResourceName { get; set; }
        public string OperationName { get; set; }
        public string SwaggerOperationId { get; set; }
        public string SdkOperationId { get; set; }
        public string Description { get; set; }
        public string FullUniqueName { get; set; }

        public string OperationNameWithParameters { get; set; }
        public string OperationNameWithScopeAndParameters { get; set; }

        /// <summary>
        /// Include parameter in or not in propertybag;
        /// </summary>
        public List<ParameterDesc> OperationMethodParameters { get; set; } = new List<ParameterDesc>();
        public ParameterDesc PropertyBagParameter;
        public List<CodeSegmentDesc> CodeSegments { get; set; } = new List<CodeSegmentDesc>();

        public AzureResourceType OperationProviderAzureResourceType { get; set; }
        public string OperationProviderType { get; set; }

        public string RequestPath { get; set; }
        public string ApiVersion { get; set; }

        public SchemaStore SchemaStore { get; set; } = SchemaStore.Current;

        public ApiDesc()
        {
        }


        public void AddCodeSegment(CodeSegmentDesc newSegment)
        {
            this.CodeSegments.Add(newSegment);
        }

        public string ToCode(bool applyUsing, bool useSuggestedName)
        {
            string newLine = "\r\n";
            var usingList = this.CodeSegments.SelectMany(s => s.UsingNamespaces)
                    .Distinct(StringComparer.Create(CultureInfo.InvariantCulture, true))
                    .OrderBy(s => s.ToLower().StartsWith("system") ? "_" + s : s).ToList();

            var usingsCode = String.Join("", usingList.Select(s => $"using {s};{newLine}"));
            var code = string.Join(newLine, this.CodeSegments.Select(s => useSuggestedName? s.GetCodeWithSuggestedName() : s.Code));
            if (applyUsing)
            {
                foreach (var u in usingList.Reverse<string>())
                {
                    code = code.Replace($"global::{u}.", "");
                }
            }
            return usingsCode + newLine + code;
        }

        //public void RefreshSchema(SchemaStore store)
        //{
        //    this.SchemaObjects = store.ObjectSchemas.Values.ToList();
        //    this.SchemaEnums = store.EnumSchemas.Values.ToList();
        //    this.SchemaNones = store.NoneSchemas.Values.ToList();
        //}

        public ApiHint GetApiHint(ExampleDesc ex, SchemaStore schemaStore = null)
        {
            ApiHint data = new ApiHint();
            foreach (var item in CodeSegments.SelectMany(cs => cs.GetFieldHints($"{this.ServiceName}.{this.ResourceName}.{this.OperationName}", ex, schemaStore)))
            {
                data.AppendFieldHint(item);
            }
            return data;
        }

        public string ToYaml()
        {
            return this.SerializeToYaml();
        }
    }
}

File: src/AutoRest.SdkExplorer.TypeScript/src/Model/Code/ApiDesc.ts
import { isStringEqualCaseInsensitive } from "../../Utils/utils";
import { AzureResourceType } from "../Azure/AzureResourceType";
import { ExampleDesc } from "../Example/ExampleDesc";
import { SchemaStore } from "../Schema/SchemaStore";
import { CodeSegmentDesc } from "./CodeSegmentDesc";
import { ParameterDesc } from "./ParameterDesc";

export class ApiDesc implements AutoRest.SdkExplorer.Interface.ApiDesc {
    language!: string;
    sdkPackageName!: string;
    sdkPackageVersion!: string;
    explorerCodeGenVersion!: string;
    generatedTimestamp!: string;
    dependencies!: string[];
    serviceName!: string;
    resourceName!: string;
    operationName!: string;
    swaggerOperationId!: string;
    sdkOperationId!: string;
    description!: string;
    apiDescription!: string;
    aiDescription!: string;
    fullUniqueName!: string;
    encodedFunctionName!: string;
    operationNameWithParameters!: string;
    operationNameWithScopeAndParameters!: string;
    operationProviderAzureResourceType?: AzureResourceType;
    operationProviderType?: string;
    apiVersion!: string;
    requestPath!: string;
    operationMethodParameters: ParameterDesc[] = [];
    codeSegments: CodeSegmentDesc[] = [];
    schemaStore?: SchemaStore;
    embeddingText?: string;
    embeddingVector?: string;


    private _examples: ExampleDesc[] = [];
    get exampleInitialized(): boolean { return this._examples.length > 0; }
    get examples(): ExampleDesc[] { return this._examples; }

    private _isSchemaLoaded: boolean = false;
    get isSchemaLoaded(): boolean { return this._isSchemaLoaded; }

    constructor(data: AutoRest.SdkExplorer.Interface.ApiDesc) {
        this.refresh(data);
    }

    public refresh(data: AutoRest.SdkExplorer.Interface.ApiDesc) {
        this.language = data.language!;
        this.sdkPackageName = data.sdkPackageName!;
        this.sdkPackageVersion = data.sdkPackageVersion!;
        this.explorerCodeGenVersion = data.explorerCodeGenVersion!;
        this.generatedTimestamp = data.generatedTimestamp!;
        this.dependencies = data.dependencies!;
        this.serviceName = data.serviceName!;
        this.resourceName = data.resourceName!;
        this.operationName = data.operationName!;
        this.swaggerOperationId = data.swaggerOperationId!;
        this.sdkOperationId = data.sdkOperationId!;
        this.description = data.description!;
        this.apiDescription = data.apiDescription ?? "";
        this.aiDescription = data.aiDescription ?? "";
        this.fullUniqueName = data.fullUniqueName!;
        this.encodedFunctionName = data.encodedFunctionName!;
        this.operationNameWithParameters = data.operationNameWithParameters!;
        this.operationNameWithScopeAndParameters = data.operationNameWithScopeAndParameters!;
        this.operationProviderAzureResourceType = data.operationProviderAzureResourceType ? new AzureResourceType(data.operationProviderAzureResourceType) : undefined;
        this.operationProviderType = data.operationProviderType;
        this.apiVersion = data.apiVersion!;
        this.requestPath = data.requestPath!;
        this.operationMethodParameters = data.operationMethodParameters?.map(d => new ParameterDesc(d)) ?? [];
        this.codeSegments = data.codeSegments?.map(d => new CodeSegmentDesc(d)) ?? [];
        this.schemaStore = data.schemaStore ? new SchemaStore(data.schemaStore) : undefined;
        this.embeddingText = data.embeddingText;
        this.embeddingVector = data.embeddingVector;
        this.loadSchema();
    }

    private loadSchema() {
        const ss = this.schemaStore;
        if (ss && ss.schemaCount > 0) {
            this.operationMethodParameters?.forEach(p => p.loadSchema(ss));
            this.allParameters.forEach(p => p.loadSchema(ss));
            this._isSchemaLoaded = true;
        }
    }

    public get allParameters(): ParameterDesc[] {
        return this.codeSegments?.flatMap((v) => {
            let r: ParameterDesc[] = v.parameters ?? [];
            return r;
        }) ?? [];
    }

    public get isCreateOperation(): boolean {
        return isStringEqualCaseInsensitive(this.operationName, "createorupdate") ||
            isStringEqualCaseInsensitive(this.operationName, "create");
    }

    public get isGetOperation(): boolean {
        return isStringEqualCaseInsensitive(this.operationName, "get");
    }
}
